File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.py
from collections import defaultdict, deque
import sys

def solution(p):
    sys.setrecursionlimit(10**6)
    
    return do(p)
    
    

def do(p):
    if not p:
        return p
    
    d = defaultdict(int)
    
    index = 0
    while d["("] == 0 or d["("] != d[")"]:
        d[p[index]] += 1
        index += 1
    
    u, v = p[:index], p[index:]

    if is_correct(u):
        return u  + do(v)
    return "(" + do(v) + ")" + u[1:-1].replace("(", "[").replace(")", "(").replace("[", ")")
        

        
        
def is_correct(u):
    stack = 0
    
    for s in u:
        if s == "(":
            stack += 1
        else:
            if stack == 0:
                return False
            stack -= 1
    
    if stack > 0:
        return False
    return True

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.java
import java.util.*;

class Solution {
    public String solution(String p) {
        if (p.isEmpty() || isCorrect(p)) {
            return p;
        }
        
        int cnt = 0;
        String u = "";
        String v = "";
        
        for(int i = 0; i < p.length(); i ++){
            if(p.charAt(i) == '(') {
                cnt++;
            }
            else {
              cnt--;  
            } 

            if(cnt == 0){
                u = p.substring(0, i + 1);
                v = p.substring(i + 1);
                break;
            }
        }
        
        if (isCorrect(u)) {
            return u + solution(v);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < u.length() - 1; i++) {
            if (u.charAt(i) == '(') {
                sb.append(")");
            } else {
                sb.append("(");
            }
        }
        
        return "(" + solution(v) + ")" + sb.toString();
    }
    
    public boolean isCorrect(String u) {
        int stack = 0;
        
        for (char c : u.toCharArray()) {
            if (c == '(') {
                stack++;
            } else {
                if (stack == 0) {
                    return false;
                }
                stack--;
            }
        }
        
        return stack == 0;
    }
}